server/src/main/java/org/opensearch/search/optimization/ranges/OptimizationContext.java

```diff
@@ -0,0 +1,410 @@
+    private boolean preparedAtShardLevel = false;
+
+    private final AggregatorBridge aggregatorBridge;
+    int maxAggRewriteFilters;
+    String shardId;
+
+    private Ranges ranges;
+
+    // debug info related fields
+    private int leaf;
+    private int inner;
+    private int segments;
+    private int optimizedSegments;
+
+    public OptimizationContext(AggregatorBridge aggregatorBridge) {
+        this.aggregatorBridge = aggregatorBridge;
+    }
+
+    public boolean canOptimize(final Object parent, final int subAggLength, SearchContext context) {
+        if (context.maxAggRewriteFilters() == 0) return false;
+
+        if (parent != null || subAggLength != 0) return false;
+
+        boolean rewriteable = aggregatorBridge.canOptimize();
+        this.rewriteable = rewriteable;
+        if (rewriteable) {
+            aggregatorBridge.setOptimizationContext(this);
+            this.maxAggRewriteFilters = context.maxAggRewriteFilters();
+            this.shardId = context.indexShard().shardId().toString();
+        }
+        logger.debug("Fast filter rewriteable: {} for shard {}", rewriteable, shardId);
+        return rewriteable;
+    }
+
+    public void prepare() throws IOException {
+        assert ranges == null : "Ranges should only be built once at shard level, but they are already built";
+        this.aggregatorBridge.buildRanges();
+        if (ranges != null) {
+            preparedAtShardLevel = true;
+        }
+    }
+
+    private Ranges prepare(LeafReaderContext leaf) throws IOException {
+        return this.aggregatorBridge.buildRanges(leaf);
+    }
+
+    void setRanges(Ranges ranges) {
+        this.ranges = ranges;
+    }
+
+    /**
+     * Try to populate the bucket doc counts for aggregation
+     * <p>
+     * Usage: invoked at segment level — in getLeafCollector of aggregator
+     *
+     * @param incrementDocCount consume the doc_count results for certain ordinal
+     */
+    public boolean tryOptimize(final LeafReaderContext leafCtx, final BiConsumer<Long, Long> incrementDocCount) throws IOException {
+        segments++;
+        if (!rewriteable) {
+            return false;
+        }
+
+        if (leafCtx.reader().hasDeletions()) return false;
+
+        PointValues values = leafCtx.reader().getPointValues(aggregatorBridge.fieldType.name());
+        if (values == null) return false;
+        // only proceed if every document corresponds to exactly one point
+        if (values.getDocCount() != values.size()) return false;
+
+        NumericDocValues docCountValues = DocValues.getNumeric(leafCtx.reader(), DocCountFieldMapper.NAME);
+        if (docCountValues.nextDoc() != NO_MORE_DOCS) {
+            logger.debug(
+                "Shard {} segment {} has at least one document with _doc_count field, skip fast filter optimization",
+                shardId,
+                leafCtx.ord
+            );
+            return false;
+        }
+
+        Ranges ranges = prepareFromSegment(leafCtx);
```

Comment:
`Ranges ranges` to `this.ranges`?

Reply:
this.ranges actually is used to represent ranges built at shard level and will be re-used for each segment.
If this.ranges is null, each segment will built its ranges if it's functionally match-all, since these are not the same for every segments, so using a local variable